sklearn/svm/src/liblinear/liblinear_helper.c

```diff
@@ -63,11 +74,14 @@ static struct feature_node **dense_to_sparse(double *x, int n_samples,
 
 
 /*
- * Convert scipy.sparse.csr to libsvm's sparse data structure
+ * Convert scipy.sparse.csr to liblinear's sparse data structure
  */
-static struct feature_node **csr_to_sparse(double *values, int *indices,
-        int *indptr, int n_samples, int n_features, int n_nonzero, double bias)
+static struct feature_node **csr_to_sparse(void *x, int double_precision,
```

Comment:
maybe use  a `char*` for consistency. Also, since (np.array).data gives a char* we can be specific

Reply:
This is exactly the kind of thing `void*` is for: In no case do we want to read the array data as chars because that would result in gibberish, and it would be a mistake to do so. `void*` prevents data access until it has been determined whether the array is an array of floats or an array of doubles.

Nonetheless, the machine code generated is the same whether x is defined as `void*` or `char*`. So if you want to give up the protection against accidentally accessing the array data as chars, it won't hurt anything to use `char*`. Let me know and I'll make the change.